woocommerce/woocommerce
# Add DB table structure for custom order tables.

### All Submissions:


### Changes proposed in this Pull Request:

This adds the base table structure for the custom order table as described in [the developer blog post](https://developer.woocommerce.com/2022/01/17/the-plan-for-the-woocommerce-custom-order-table/). The tool in _Settings > Tools_ has now two functions: "Create the tables" and "Delete the tables" (so no more references to migration, syncrhonization, etc; these will be handled by/displayed in settings once implemented).

Additionally, a new `DatabaseUtil` class is added which contains (amongst other stuff) a refactored version of the `verify_db` tool.

Closes #31612.

### How to test the changes in this Pull Request:

1. Enable the custom orders tables feature by adding the code snippet: 

```php
wc_get_container()
  ->get('Automattic\WooCommerce\Internal\DataStores\Orders\CustomOrdersTableController')
  ->show_feature();
```

2. Go to WooCommerce > Settings > Advanced, you'll see a new "Custom data stores" section with a message stating that there's nothing to configure yet:

![image](https://user-images.githubusercontent.com/937723/153186154-ca1f8fd6-29a8-48c4-8a18-7c4d8f44d249.png)

2. In a separate browser tab go to WooCommerce > Status > Tools, find a tool named "Create the custom orders tables" and run it, it will then become a "Delete the custom orders tables" tool.

Before running:

![image](https://user-images.githubusercontent.com/937723/153186375-6e1545ec-903a-41b5-ad55-a86688670a9f.png)

After running:

![image](https://user-images.githubusercontent.com/937723/153188293-739e2234-be4a-4183-a5c7-e9c1d4c231e6.png)

![image](https://user-images.githubusercontent.com/937723/153186885-10ee0af6-12cb-40f9-9443-87ebaa4d35b1.png)

3. Verify that the database tables have been created (`wp_wc_orders`, `wp_wc_order_addresses` and `wp_wc_order_operational_data`).

4. Go back to the settings tab, reload and you'll now see a setting that allows you to enable the table usage. Enable it and save.

![image](https://user-images.githubusercontent.com/937723/153365717-77260ae1-bcef-4040-84aa-7179d7a219ef.png)

5. Go back to the tools page, DON'T reload it and click "Save". You should get an error and you'll see that the "Delete" button is now disabled.

![image](https://user-images.githubusercontent.com/937723/153188759-72c8e528-332c-483b-aa11-06aaf9ba01fe.png)

![image](https://user-images.githubusercontent.com/937723/153188906-c8672e28-12e9-43c3-a9d6-d2a8f8a3c49a.png)

6. Go back to settings, disable the table usage and save.

7. Go back to the tools page, reload it, the "Delete" button should now be enabled. Run the tool.

8. Verify that the tables have been indeed deleted, and that reloading the settings page shows the "Create the tables firsts..." message again.

### Other information:


### Changelog entry

> Enhancement - Add DB table structure for custom order tables.

### FOR PR REVIEWER ONLY:
